apps/aesophia/test/contracts/erc20_token.aes

```diff
@@ -0,0 +1,97 @@
+contract ERC20Token =
+  record state = {
+    totalSupply : int,
+    decimals : int,
+    name : string,
+    symbol : string,
+    balances : map(address, int),
+    allowed : map(address, map(address,int)),
+    // Logs, remove when native Events are there
+    transfer_log : list((address,address,int)),
+    approval_log : list((address,address,int))}
+
+  // init(100000000, 10, "Token Name", "TKN")
+  public stateful function init(_totalSupply : int, _decimals : int, _name : string, _symbol : string ) = {
+    totalSupply = _totalSupply,
+    decimals = _decimals,
+    name = _name,
+    symbol = _symbol,
+    balances = {[Call.caller] = _totalSupply }, // creator gets all Tokens
+    allowed = {},
+    // Logs, remove when native Events are there
+    transfer_log = [],
+    approval_log = []}
+
+  public stateful function totalSupply() : int = state.totalSupply
+  public stateful function decimals() : int = state.decimals
+  public stateful function name() : string = state.name
+  public stateful function symbol() : string = state.symbol
+
+  public stateful function balanceOf(tokenOwner : address ) : int = lookupDefault(tokenOwner, state.balances, 0)
+
+  public stateful function transfer(to : address, tokens : int) =
+    put( state{balances[Call.caller] = sub(state.balances[Call.caller], tokens) })
+    put( state{balances[to] = add(lookupDefault(to, state.balances, 0), tokens) })
+    transferEvent(Call.caller, to, tokens)
+    true
+
+  public stateful function approve(spender : address, tokens : int) =
+    // allowed[Call.caller] field must have a value!
+    ensure_allowed(Call.caller)
```

Comment:
I see this contains a put state, then next line is a put state too. I see other functions contain multiple puts. Any reasons for not having only one put, preparing state as a variable before that?

* Having only one put state at the end would seem more readable. Though debatable.
* The intended comparative gas cost between multiple [put](https://github.com/aeternity/epoch/blob/85f5050fbdf4369385ba33a8b51ff8dfed044017/apps/aesophia/src/aeso_icode_to_asm.erl#L367-L370)s and using variable may need to be considered.

Reply:
`put` is essentially free. The only thing it does is write a single word to address 0.

Reply:
The reason for the ensure_allowed is that the next put accesses maps at two levels `allowed[from][Call.caller] = ...` and sophia requires the outermost map to exist. This is what ensured_allowed does.

Reply:
Ok.